### PR TITLE
Allow creating and editing system webhooks.

### LINKED ISSUE #23139: System Hooks API confused with Default Hooks API
### Description

`GET /admin/hooks` lists hooks with `is_system_webhook = true`, but `POST /admin/hooks` creates them without. There does not currently seem to be a way to create a system webhook from the API -- only default webhooks; meanwhile there is no way to list default webhooks from the API.

I see that `/admin/hooks` was only added less than a month ago with https://github.com/go-gitea/gitea/pull/14537. It's an niche but extremely helpful feature for what I'm trying to do with Gitea, so I want to help get it working.

### Gitea Version

1.19.0+dev-586-g10cdcb9ea

### Can you reproduce the bug on the Gitea demo site?

No -- but only because I don't have admin rights there

### Log Gist

https://gist.github.com/kousu/1f0be0c81dce500caef0b0af3e985196

### Screenshots

Under http://localhost:3000/admin/hooks I start with an empty slate

Screenshot 2023-02-25 at 01-27-23 Gitea Git with a cup of tea

I go to http://localhost:3000/api/swagger and log in:

Screenshot 2023-02-25 at 01-26-31 Gitea API

http://localhost:3000/api/swagger#/admin/adminGetHook gives an empty list

Screenshot 2023-02-25 at 01-30-46 Gitea API

I create a hook:

Screenshot 2023-02-25 at 01-30-26 Gitea API

Back on the *UI* at http://localhost:3000/admin/hooks I can see it, but it's a Default Hook:

Screenshot 2023-02-25 at 01-31-25 Gitea Git with a cup of tea

But using the API gets me an empty list still:

Screenshot 2023-02-25 at 01-32-07 Gitea API

If I make a system webhook *with the UI*

Screenshot 2023-02-25 at 01-33-16 Gitea Git with a cup of tea

it adds to the UI

Screenshot 2023-02-25 at 01-33-31 Gitea Git with a cup of tea

and the API can see it:

Screenshot 2023-02-25 at 01-33-59 Gitea API

If I add another Default Webhook

Screenshot 2023-02-25 at 01-34-33 Gitea Git with a cup of tea

The API **cannot** see it

Screenshot 2023-02-25 at 01-34-58 Gitea API

...unless I specifically ask for hook 20

Screenshot 2023-02-25 at 01-35-44 Gitea API

19 also works, but there's nothing to distinguish that it is a system hook:

Screenshot 2023-02-25 at 01-36-12 Gitea API

### Git Version

2.34.1

### Operating System

Ubuntu 22.04

### How are you running Gitea?

```
user@dev:~/src/gitea$ TAGS="bindata sqlite sqlite_unlock_notify" make build
user@dev:~/src/gitea$ ./gitea
```

### Database

SQLite

### PR DESCRIPTION
**Pull request by kousu**
Fixes https://github.com/go-gitea/gitea/issues/23139, hopefully.

Also, allow editing (but not viewing) the _secret_ associated with each hook.

### CODE REVIEW

**Comment by kousu** [comment]
This code is not elegant. It fixes the problems I ran into, but I'm sure it does it suboptimally. For one thing, `AddSystemHook` should be renamed. And maybe there should be separate `/admin/hooks/system`, `/admin/hooks/default` routes.

**Comment by kousu** [comment]
Thank you for taking the time to curate this, @yardenshoham; I don't think this is an enhancement though, it is a bug fix to a feature that went in to 1.19 and isn't usable at the moment. I was hoping to get this fixed in 1.19.

**Comment by yardenshoham** [comment]
Labels make sense now?

**Comment by kousu** [comment]
Yes! Thank you :) :)

**Comment by kousu** [comment]
I want to be able to edit `secret` because I want to automate deployments with terraform/Docker/etc -- that's the main reason I'm using the API and not just giving install instructions for an admin to follow.

If I can edit a secret, I can do:

- `GET /admin/hooks`
- if a hook pointing at the expected URL exists: `PATCH /admin/hooks/{id}`
- else: `POST /admin/hooks`

If I can't edit a secret, I can instead do:

- `GET /admin/hooks`
- if a hook pointing at the expected URL exists: `DELETE /admin/hooks/{id}`
- `POST /admin/hooks`

This will wipe out logs of previous attempts on the hook, which is a little unfortunate, and requires 3 APIs calls then where the other needs 2. It's not the end of the world if you don't want to allow editing the secret; I can think of arguments in favour of that too, like making hooks symmetrical with tokens which also don't allow editing after their creation.

File: models/webhook/webhook_system.go [hunk 1]
```diff
@@ -46,6 +46,19 @@ func GetSystemWebhooks(ctx context.Context, isActive util.OptionalBool) ([]*Webh
...
+	if isActive.IsNone() {
+		return webhooks, db.GetEngine(ctx).
+			Where("repo_id=? AND org_id=?", 0, 0).
+			Find(&webhooks)
+	}
+	return webhooks, db.GetEngine(ctx).
+		Where("repo_id=? AND org_id=? AND is_active = ?", 0, 0, isActive.IsTrue()).
+		Find(&webhooks)
+}
+
```
Comment (kousu) [inline models/webhook/webhook_system.go]: There's improvements to be made here. For example, if this exists, do we really still need `GetSystemWebhooks()`? Would you ever use both?

**Comment by kousu** [comment]
Hi! I hope you're all doing well with working on Actions. Is there any attention available for looking this patch over?

It's a pretty obvious bug and what I think is a straightforward fix.

File: modules/structs/hook.go [hunk 2]
```diff
@@ -24,6 +24,7 @@ type Hook struct {
 	Events              []string          `json:"events"`
 	AuthorizationHeader string            `json:"authorization_header"`
 	Active              bool              `json:"active"`
+	IsSystemWebhook     bool              `json:"is_system_webhook"`
```
Comment (lunny) [inline modules/structs/hook.go]: Although the database stored as the name, but maybe we can have a better name like `IsSystemOrDefault`? Or even `ScopeType`?

File: routers/api/v1/admin/hooks.go [hunk 3]
```diff
@@ -36,7 +36,7 @@ func ListHooks(ctx *context.APIContext) {
 	//   "200":
 	//     "$ref": "#/responses/HookList"
 
-	sysHooks, err := webhook.GetSystemWebhooks(ctx, util.OptionalBoolNone)
+	sysHooks, err := webhook.GetSystemOrDefaultWebhooks(ctx, util.OptionalBoolNone)
```
Comment (lunny) [inline routers/api/v1/admin/hooks.go]: I don't think we should change the behaviour? Maybe we should have parameter or we could have another endpoint for default webhooks?

**Comment by kousu** [comment]
I'm still interested in this. I started to rewrite some parts to adopt Lunny's feedback but work got in the way before I got my rewrite building. I do want it to come though!

File: modules/structs/hook.go (hunk 2 above)
Comment (kousu) [reply modules/structs/hook.go]: I've gotten around to this again finally.

I decided to replace `IsSystemWebhook` by a path parameter `hookType`:

https://github.com/go-gitea/gitea/blob/5384c1d877a99e8fb3d5be148643490ad2afd68d/routers/api/v1/api.go#L1384-L1390

https://github.com/go-gitea/gitea/blob/5384c1d877a99e8fb3d5be148643490ad2afd68d/routers/api/v1/admin/hooks.go#L24

https://github.com/go-gitea/gitea/blob/5384c1d877a99e8fb3d5be148643490ad2afd68d/routers/api/v1/admin/hooks.go#L62

https://github.com/go-gitea/gitea/blob/5384c1d877a99e8fb3d5be148643490ad2afd68d/routers/api/v1/admin/hooks.go#L107

https://github.com/go-gitea/gitea/blob/5384c1d877a99e8fb3d5be148643490ad2afd68d/routers/api/v1/admin/hooks.go#L139

https://github.com/go-gitea/gitea/blob/5384c1d877a99e8fb3d5be148643490ad2afd68d/routers/api/v1/admin/hooks.go#L178


Doing this ballooned the PR, but it makes the API more friendly:

Screenshot 2023-07-01 at 23-41-57 Gitea API

and it also matches how the admin web UI works, where the two links on `/admin/system-hooks/`

Screenshot 2023-07-01 at 23-43-17 Gitea Git with a cup of tea

go to `/admin/system-hooks/{kind}/new` and `http://localhost:3000/admin/default-hooks/{kind}/new` respectively.

What do you think?

**Comment by kousu** [comment]
I've redone this to adopt https://github.com/go-gitea/gitea/pull/23142#discussion_r1185678106 into it, and then cleaned up the API/UI endpoints to match each other. Since that was a pretty extensive change, I've resubmitted this as https://github.com/go-gitea/gitea/pull/26418.

**Pull request closed without merging**